File: packages/ref-id/src/validators.ts
// SPDX-License-Identifier: Apache-2.0
//
// The one registry that maps a NAME the spec uses to behaviour: the owning validators (by
// `dispatch.<type>.validator`), the ways a delegated string is formed (by `dispatch.<type>.delegate`),
// the refinement range constraints (by `refinements.<key>.range`), and the policies this package
// implements (`unknownKey`, `repeatedKey`). A validator or delegate name the spec uses and this file
// does not know degrades to `uncovered`; a range or policy name it does not know is a version
// mismatch and is refused loudly, because silently skipping a constraint is the drift the guardrail
// forbids.

import { PackageURL } from "packageurl-js"
import { FIELD, pattern } from "./grammar.ts"
import { SpecVersionError, type RefIdSpec } from "./spec.ts"

export type Validation = { ok: true; canonical?: string } | { ok: false }

type DispatchEntry = RefIdSpec["dispatch"][string]

const validators: Record<string, (spec: RefIdSpec, entry: DispatchEntry, delegated: string) => Validation> = {
  "package-url": (_spec, _entry, delegated) => {
    const { base, subpath } = splitSubpath(delegated)
    try {
      const purl = PackageURL.fromString(base)
      if (subpath === undefined) return { ok: true, canonical: purl.toString() }
      const withPath = new PackageURL(purl.type, purl.namespace, purl.name, purl.version, purl.qualifiers, subpath)
      return { ok: true, canonical: withPath.toString() }
    } catch {
      return { ok: false }
    }
  },
  "declared-name": (spec, entry, delegated) => ({ ok: pattern(spec, entry.pattern ?? "").test(delegated) }),
  // A number whose shape is right and whose check digit is wrong is a typo, not an identifier — and it
  // is the one error a pattern cannot see. Both registered schemes reduce to the same weighted sum, so
  // one function serves them; neither adds a dependency, which the portability guardrail requires.
  "check-digit": (spec, entry, delegated) => ({
    ok: pattern(spec, entry.pattern ?? "").test(delegated) && checkDigitHolds(delegated),
  }),
}

/**
 * Where a Package URL locator stops being the package and starts being a path inside it.
 *
 * An `@` that opens a segment — preceded by `/`, or first in the string — belongs to a namespace and is
 * part of the name. An `@` inside a segment closes the name: the version runs from it to the next `/`,
 * and whatever follows that `/` is the subpath. Without a version there is no marker at all, so
 * `npm/a/b/c` stays a namespaced package and carries no subpath — a file inside a corpus nobody
 * versioned is named through `folder`, whose locator needs no such marker.
 *
 * The subpath leaves here as the Package URL's own `#subpath` component, which is what that component
 * means; the scheme's `#` is the declared name one level below the file and has no purl equivalent.
 */
function splitSubpath(delegated: string): { base: string; subpath?: string } {
  for (let index = 1; index < delegated.length; index += 1) {
    if (delegated[index] !== "@" || delegated[index - 1] === "/") continue
    const slash = delegated.indexOf("/", index)
    return slash < 0 ? { base: delegated } : { base: delegated.slice(0, slash), subpath: delegated.slice(slash + 1) }
  }
  return { base: delegated }
}

/**
 * The check digit of a registered article number.
 *
 * A ten-character book number weights its digits 10..1 and is correct when the sum is divisible by
 * eleven, which is why its last character may be `X` for the value ten. Every other length is a GS1
 * trade item number: the digits before the last are weighted 3 and 1 alternately from the right, and the
 * last is whatever brings the total up to a multiple of ten.
 */
function checkDigitHolds(value: string): boolean {
  if (value.length === 10) {
    const weighted = [...value].reduce((sum, char, index) => sum + (char === "X" ? 10 : Number(char)) * (10 - index), 0)
    return weighted % 11 === 0
  }
  const digits = [...value].map(Number)
  const declared = digits.at(-1)
  const weighted = digits
    .slice(0, -1)
    .reverse()
    .reduce((sum, digit, index) => sum + digit * (index % 2 === 0 ? 3 : 1), 0)
  return (10 - (weighted % 10)) % 10 === declared
}

/** How `dispatch.<type>.delegate` forms the delegated string, and whether a producer may pass the format's own scheme back in. */
const delegators: Record<string, { form: (type: string, locator: string) => string; foldsType: boolean }> = {
  "type-prefixed": { form: (type, locator) => `${type}${FIELD}${locator}`, foldsType: true },
  verbatim: { form: (_type, locator) => locator, foldsType: false },
}

/** The constraints a refinement may declare beyond its pattern, by the name `refinements.<key>.range` uses. */
const ranges: Record<string, (value: string, boundSeparator: string) => boolean> = {
  ascending: (value, boundSeparator) => {
    const bounds = value.split(boundSeparator).map(Number)
    return bounds.length < 2 || (bounds[0] ?? 0) <= (bounds[1] ?? 0)
  },
}

/** The policies this package implements, by the name the spec declares them under. */
const policies = { unknownKey: "carry-through", repeatedKey: "malformed" } as const

const policiesChecked = new WeakSet<RefIdSpec>()

/** Refuses a spec whose declared policies or range names are ones this package does not implement. */
export function assertImplemented(spec: RefIdSpec): void {
  if (policiesChecked.has(spec)) {
    return
  }
  const declared: [string, string | undefined][] = [
    ["grammar.state.unknownKey", spec.grammar.state.unknownKey],
    ["unknownRefinement", spec.unknownRefinement],
    ["grammar.state.repeatedKey", spec.grammar.state.repeatedKey],
    ["grammar.fragment.repeatedKey", spec.grammar.fragment.repeatedKey],
  ]
  for (const [field, value] of declared) {
    const implemented = field.endsWith("repeatedKey") ? policies.repeatedKey : policies.unknownKey
    if (value !== implemented) {
      throw new SpecVersionError(`spec.${field} declares ${JSON.stringify(value)}; this package implements only ${JSON.stringify(implemented)}`)
    }
  }
  for (const [key, refinement] of Object.entries(spec.refinements)) {
    if (refinement.range !== undefined && !Object.hasOwn(ranges, refinement.range)) {
      throw new SpecVersionError(`spec.refinements.${key}.range declares ${JSON.stringify(refinement.range)}, which this package does not implement`)
    }
  }
  policiesChecked.add(spec)
}

/** The string handed to the validator, or undefined when the type is not dispatched or its delegate mode is unknown here. */
export function delegatedString(spec: RefIdSpec, type: string, locator: string): string | undefined {
  const entry = spec.dispatch[type]
  const delegator = entry ? delegators[entry.delegate] : undefined
  return delegator ? delegator.form(type, locator) : undefined
}

/** True when a producer may pass the intact format string (`pkg:npm/x`) as the locator of this type. */
export function foldsType(spec: RefIdSpec, type: string): boolean {
  const entry = spec.dispatch[type]
  return entry ? (delegators[entry.delegate]?.foldsType ?? false) : false
}

/** Runs the owning validator, or undefined when the spec names one this package does not implement. */
export function validateLocator(spec: RefIdSpec, entry: DispatchEntry, delegated: string): Validation | undefined {
  const validator = validators[entry.validator]
  return validator ? validator(spec, entry, delegated) : undefined
}

/** True when the value satisfies the refinement's declared range constraint (or it declares none). */
export function rangeHolds(refinement: RefIdSpec["refinements"][string], value: string): boolean {
  if (refinement.range === undefined) {
    return true
  }
  const check = ranges[refinement.range]
  return check ? check(value, refinement.boundSeparator ?? "") : true
}

File: crates/ref-id/src/validators.rs
// SPDX-License-Identifier: Apache-2.0
//
// The one registry that maps a NAME the spec uses to behaviour: the owning validators (by
// `dispatch.<type>.validator`), the ways a delegated string is formed (by `dispatch.<type>.delegate`),
// the refinement range constraints (by `refinements.<key>.range`) and the policies this crate
// implements. A validator or delegate name the spec uses and this file does not know degrades to
// `uncovered`; a range or policy name it does not know is a version mismatch, refused loudly.

use crate::grammar::{Grammar, FIELD};
use crate::spec::Spec;
use crate::types::RefIdError;
use serde_json::Value;
use std::str::FromStr;

pub(crate) struct Validation {
    pub ok: bool,
    pub canonical: Option<String>,
}

/// Where a Package URL locator stops being the package and starts being a path inside it.
///
/// An `@` that opens a segment — preceded by `/`, or first in the string — belongs to a namespace and is
/// part of the name. An `@` inside a segment closes the name: the version runs from it to the next `/`,
/// and whatever follows that `/` is the subpath. Without a version there is no marker at all, so
/// `npm/a/b/c` stays a namespaced package and carries no subpath — a file inside a corpus nobody
/// versioned is named through `folder`.
///
/// The subpath leaves here as the Package URL's own `#subpath` component, which is what that component
/// means; the scheme's `#` is the declared name one level below the file and has no purl equivalent.
/// Ported from `packages/ref-id/src/validators.ts:52` (`splitSubpath`).
fn split_subpath(delegated: &str) -> (&str, Option<&str>) {
    let chars: Vec<(usize, char)> = delegated.char_indices().collect();
    for i in 1..chars.len() {
        let (byte_idx, c) = chars[i];
        if c != '@' || chars[i - 1].1 == '/' {
            continue;
        }
        return match delegated[byte_idx..].find('/') {
            None => (delegated, None),
            Some(rel) => {
                let slash = byte_idx + rel;
                (&delegated[..slash], Some(&delegated[slash + 1..]))
            }
        };
    }
    (delegated, None)
}

fn package_url(_spec: &Spec, _entry: &Value, delegated: &str) -> Result<Validation, RefIdError> {
    let (base, subpath) = split_subpath(delegated);
    Ok(match packageurl::PackageUrl::from_str(base) {
        Ok(mut purl) => {
            if let Some(path) = subpath {
                if purl.with_subpath(path).is_err() {
                    return Ok(Validation { ok: false, canonical: None });
                }
            }
            Validation { ok: true, canonical: Some(purl.to_string()) }
        }
        Err(_) => Validation { ok: false, canonical: None },
    })
}

fn declared_name(spec: &Spec, entry: &Value, delegated: &str) -> Result<Validation, RefIdError> {
    let pattern = entry.get("pattern").and_then(Value::as_str).unwrap_or("");
    Ok(Validation { ok: Grammar::of(spec)?.matches(spec, pattern, delegated)?, canonical: None })
}

/// The check digit of a registered article number.
///
/// A ten-character book number weights its digits 10..1 and is correct when the sum is divisible by
/// eleven, which is why its last character may be `X` for the value ten. Every other length is a GS1
/// trade item number: the digits before the last are weighted 3 and 1 alternately from the right, and the
/// last is whatever brings the total up to a multiple of ten. Ported from
/// `packages/ref-id/src/validators.ts:44` (`checkDigitHolds`).
fn check_digit_holds(value: &str) -> bool {
    let chars: Vec<char> = value.chars().collect();
    if chars.len() == 10 {
        let weighted: u64 = chars
            .iter()
            .enumerate()
            .map(|(index, &c)| {
                let digit = if c == 'X' { 10 } else { c.to_digit(10).unwrap_or(0) as u64 };
                digit * (10 - index as u64)
            })
            .sum();
        return weighted % 11 == 0;
    }
    let digits: Vec<u64> = chars.iter().map(|c| c.to_digit(10).unwrap_or(0) as u64).collect();
    let declared = match digits.last() {
        Some(&d) => d,
        None => return false,
    };
    let weighted: u64 = digits[..digits.len() - 1]
        .iter()
        .rev()
        .enumerate()
        .map(|(index, &digit)| digit * if index % 2 == 0 { 3 } else { 1 })
        .sum();
    (10 - (weighted % 10)) % 10 == declared
}

fn check_digit(spec: &Spec, entry: &Value, delegated: &str) -> Result<Validation, RefIdError> {
    let pattern = entry.get("pattern").and_then(Value::as_str).unwrap_or("");
    let matches = Grammar::of(spec)?.matches(spec, pattern, delegated)?;
    Ok(Validation { ok: matches && check_digit_holds(delegated), canonical: None })
}

type Validator = fn(&Spec, &Value, &str) -> Result<Validation, RefIdError>;

fn validator(name: &str) -> Option<Validator> {
    match name {
        "package-url" => Some(package_url),
        "declared-name" => Some(declared_name),
        "check-digit" => Some(check_digit),
        _ => None,
    }
}

struct Delegator {
    form: fn(&str, &str) -> String,
    folds_type: bool,
}

fn delegator(mode: &str) -> Option<Delegator> {
    match mode {
        "type-prefixed" => Some(Delegator { form: |t, l| format!("{t}{FIELD}{l}"), folds_type: true }),
        "verbatim" => Some(Delegator { form: |_, l| l.to_string(), folds_type: false }),
        _ => None,
    }
}

fn range(name: &str) -> Option<fn(&str, &str) -> bool> {
    match name {
        "ascending" => Some(|value, separator| {
            let bounds: Vec<&str> = if separator.is_empty() { vec![value] } else { value.split(separator).collect() };
            match (bounds.first().and_then(|b| b.parse::<u64>().ok()), bounds.get(1).and_then(|b| b.parse::<u64>().ok())) {
                (Some(first), Some(second)) => first <= second,
                _ => true,
            }
        }),
        _ => None,
    }
}

const UNKNOWN_KEY_POLICY: &str = "carry-through";
const REPEATED_KEY_POLICY: &str = "malformed";

/// Refuses a spec whose declared policies or range names are ones this crate does not implement.
pub(crate) fn assert_implemented(spec: &Spec) -> Result<(), RefIdError> {
    spec.implemented
        .get_or_init(|| {
            let declared = [
                ("grammar.state.unknownKey", spec.string(&["grammar", "state", "unknownKey"]), UNKNOWN_KEY_POLICY),
                ("unknownRefinement", spec.string(&["unknownRefinement"]), UNKNOWN_KEY_POLICY),
                ("grammar.state.repeatedKey", spec.string(&["grammar", "state", "repeatedKey"]), REPEATED_KEY_POLICY),
                ("grammar.fragment.repeatedKey", spec.string(&["grammar", "fragment", "repeatedKey"]), REPEATED_KEY_POLICY),
            ];
            for (field, value, implemented) in declared {
                if value != implemented {
                    return Err(RefIdError::SpecVersion(format!("spec.{field} declares {value:?}; this crate implements only {implemented:?}")));
                }
            }
            if let Some(refinements) = spec.object(&["refinements"]) {
                for (key, refinement) in refinements {
                    if let Some(name) = refinement.get("range").and_then(Value::as_str) {
                        if range(name).is_none() {
                            return Err(RefIdError::SpecVersion(format!("spec.refinements.{key}.range declares {name:?}, which this crate does not implement")));
                        }
                    }
                }
            }
            Ok(())
        })
        .clone()
}

fn dispatch_entry<'a>(spec: &'a Spec, r#type: &str) -> Option<&'a Value> {
    spec.object(&["dispatch"]).and_then(|m| m.get(r#type))
}

pub(crate) fn delegated_string(spec: &Spec, r#type: &str, locator: &str) -> Option<String> {
    let entry = dispatch_entry(spec, r#type)?;
    let mode = entry.get("delegate").and_then(Value::as_str)?;
    Some((delegator(mode)?.form)(r#type, locator))
}

pub(crate) fn folds_type(spec: &Spec, r#type: &str) -> bool {
    dispatch_entry(spec, r#type)
        .and_then(|e| e.get("delegate").and_then(Value::as_str))
        .and_then(delegator)
        .map(|d| d.folds_type)
        .unwrap_or(false)
}

/// Runs the owning validator, or None when the spec names one this crate does not implement.
pub(crate) fn validate_locator(spec: &Spec, entry: &Value, delegated: &str) -> Result<Option<Validation>, RefIdError> {
    match entry.get("validator").and_then(Value::as_str).and_then(validator) {
        Some(validate) => validate(spec, entry, delegated).map(Some),
        None => Ok(None),
    }
}

pub(crate) fn range_holds(refinement: &Value, value: &str) -> bool {
    match refinement.get("range").and_then(Value::as_str).and_then(range) {
        Some(check) => check(value, refinement.get("boundSeparator").and_then(Value::as_str).unwrap_or("")),
        None => true,
    }
}
